src/query/sql/src/planner/semantic/type_check.rs

```diff
@@ -1172,6 +1176,44 @@ impl<'a> TypeChecker<'a> {
         Ok(Box::new((scalar, data_type)))
     }
 
+    fn try_rewrite_virtual_column_cast(
+        &mut self,
+        scalar: &mut ScalarExpr,
+        target_type: &DataType,
+    ) -> bool {
+        let ScalarExpr::BoundColumnRef(column) = scalar else {
+            return false;
+        };
+        let mut guard = self.metadata.write();
+        if column.column.table_index.is_none() {
+            return false;
+        }
+        let ColumnEntry::VirtualColumn(virtual_column) = guard.column(column.column.index).clone()
+        else {
+            return false;
+        };
+        let ColumnEntry::BaseTableColumn(source_column) =
+            guard.column(virtual_column.source_column_index).clone()
+        else {
+            return false;
+        };
+        if let Some(table_ty) = Option::<TableDataType>::from(target_type) {
+            // Tips: column id from source virtual_column
+            column.column.index = guard.add_virtual_column(
```

Comment:
don't add a new virtual column, modify the exist virtual column's data_type, and also need a field to identify the data type is `cast` or `try_cast`.